src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.macOS.cs

```diff
@@ -0,0 +1,230 @@
+                ? Interop.AppleCrypto.SecKeychainCopyDefault()
+                : Interop.AppleCrypto.CreateTemporaryKeychain();
+
+            importState.Exportable = exportable;
+            importState.Persisted = persist;
+            importState.Keychain = keychain;
+        }
+
+        private static partial Pkcs12Return FromCertAndKey(CertAndKey certAndKey, ImportState importState)
+        {
+            AppleCertificatePal pal = (AppleCertificatePal)certAndKey.Cert!;
+            SafeSecKeyRefHandle? key = null;
+
+            if (certAndKey.Key is not null)
+            {
+                key = GetPrivateKey(certAndKey.Key);
+                certAndKey.Key.Dispose();
+            }
+
+            if (key is not null || importState.Persisted)
+            {
+                if (key is not null && !importState.Exportable)
+                {
+                    AppleCertificatePal newPal = AppleCertificatePal.ImportPkcs12NonExportable(
+                        pal,
+                        key,
+                        SafePasswordHandle.InvalidHandle,
+                        importState.Keychain);
+
+                    pal.Dispose();
+                    pal = newPal;
+                }
+                else
+                {
+                    AppleCertificatePal? identity = pal.MoveToKeychain(importState.Keychain, key);
+
+                    if (identity is not null)
+                    {
+                        pal.Dispose();
+                        pal = identity;
+                    }
+                }
+            }
+
+            return new Pkcs12Return(pal);
+        }
+
+        private static partial AsymmetricAlgorithm? CreateKey(string algorithm)
+        {
+            return algorithm switch
+            {
+                Oids.Rsa or Oids.RsaPss => new RSAImplementation.RSASecurityTransforms(),
+                Oids.EcPublicKey or Oids.EcDiffieHellman => new ECDsaImplementation.ECDsaSecurityTransforms(),
+                Oids.Dsa => new DSAImplementation.DSASecurityTransforms(),
+                _ => null,
+            };
+        }
+
+        internal static SafeSecKeyRefHandle? GetPrivateKey(AsymmetricAlgorithm? key)
+        {
+            if (key == null)
+            {
+                return null;
+            }
+
+            if (key is RSAImplementation.RSASecurityTransforms rsa)
+            {
+                // Convert data key to legacy CSSM key that can be imported into keychain
+                byte[] rsaPrivateKey = rsa.ExportRSAPrivateKey();
+                using (PinAndClear.Track(rsaPrivateKey))
+                {
+                    return Interop.AppleCrypto.ImportEphemeralKey(rsaPrivateKey, true);
+                }
+            }
+
+            if (key is DSAImplementation.DSASecurityTransforms dsa)
+            {
+                // DSA always uses legacy CSSM keys do no need to convert
+                return dsa.GetKeys().PrivateKey;
+            }
```

Comment:
This is, I believe, one of the reasons you are hitting an assert in the unit tests for DSA keys. For RSA and ECDSA, we export and import the keys, which gives us new `SafeSecKeyRefHandle`s with an independent lifetime from the security transform.

That does not appear to be the case with DSA. That is returning the same key handle from the transform. Later on the security transform gets disposed, and the handle gets disposed out from under it.

Reply:
I'm about to make a push of other things I've cleaned up, and this won't be fixed yet; because I'm doing the push just to get up to date on my mac where I can verify the change.